sig-security-tooling/vulnerability-mgmt/container-images.md

```diff
@@ -0,0 +1,180 @@
+# Periodic scanning for vulnerabilities in container images
+
+Report vulnerabilities in container images
+of [Kubernetes](https://github.com/kubernetes/kubernetes) repository
+
+Tracker: [Issue #101528](https://github.com/kubernetes/kubernetes/issues/101528)
+
+## Background and Prior work
+
+The process described here is tooling agnostic i.e. the process can be
+implemented using any scanner with minimal or no changes. This is also _not_ an
+endorsement of any specific tool or scanner. In order to get a working solution
+in place, [snyk](https://snyk.io/) was chosen for following reasons:
+
+1. Existing partnership between CNCF and Snyk helped procure an account that
+   allowed us to scan `kubernetes/kubernetes`
+   repo: https://github.com/kubernetes/steering/issues/206
+2. Snyk has detected vulnerabilities in transient dependencies of
+   `kubernetes/kubernetes`: https://kubernetes.slack.com/archives/CHGFYJVAN/p1595258034095300
+3. Snyk has a programmable interface which made it easier to filter out
+   licensing issues and known false positive vulnerabilities
+
+## Implementation with Snyk
+
+There are two ways to scan the Kubernetes repo for vulnerabilities in
+container images
+
+### Running the scan locally
+
+#### Step 0: Install Snyk CLI
+
+Follow these instructions to snyk cli installed on your
+machine: https://support.snyk.io/hc/en-us/articles/360003812538-Install-the-Snyk-CL
+
+#### Step 1: Authenticate
+
+##### Option A :
+
+Running command `snyk auth` takes you to snyk.io website, do signup/login/auth
+
+```
+snyk auth
+```
+
+##### Option B:
+
+Get the API token from https://app.snyk.io/account and use it
+
+```
+snyk auth XXX-XXX-XXX-XXX-XXX
+Your account has been authenticated. Snyk is now ready to be used.
+```
+
+#### Step 2: Collect list of container images 
+
+```sh
+curl -Ls https://sbom.k8s.io/$(curl -Ls https://dl.k8s.io/release/latest.txt)/release | grep 'PackageName: k8s.gcr.io/' | awk '{print $2}' > images.txt
+```
+
+#### Step 3: Run test
+
+```sh
+while read image; do snyk container test $image -d --json; done < images.txt
+```
+
+### Running the scan as part of k/k testgrid
+
+Prow job that runs every 6 hours is located
+here: https://testgrid.k8s.io/sig-security-snyk-scan#ci-kubernetes-snyk-master
+
+#### Improvements to the raw scan results
+
+Raw scan results were useful, but needed some Kubernetes specific work
+
+##### JSON output
+
+To store the json output in a file and let stdout use command line friendly
+output:
+
+```
+snyk container test --json-file-output=all-cves.json
+```
+
+#####  Fixable Vulnerabilities
+
+To output vulnerabilites able to be triaged, only upgradeable 
+or patchable results can be parsed from the output using this query:
+
+```
+cat all-cves.json | jq '{ vulnerabilities: .vulnerabilities | map(select(.isUpgradable == true or .isPatchable == true)) | select(length > 0) }' > fixable_cves.json
```

Comment:
One thing I wanted to call out is that we are filtering for only fixable vulnerabilities in our [prow job](https://github.com/kubernetes/test-infra/blob/1842eb7f9edb1e6368f26aee24c811dc02fae4b9/config/jobs/kubernetes/sig-k8s-infra/trusted/sig-security-trusted.yaml#L44-L45).  

This makes sense if our ultimate goal is to fail builds with vulnerabilities or to track our time to remediation, however should we call this out somewhere?  Maybe in the non-goals section of the README here?

Reply:
Yes good call out. The main idea is here to focus on actionable vulnerabilities i.e. the ones that can be fixed. It is definitely worth mentioning that in the non-goals though. 

Can you add the following text as part of the non-goals in the README?

```
4. **Fixing unfixable vulnerabilities**: The focus of this effort is to apply fixes to vulnerabilities for which a patch is released by owners of the vulnerable dependency or base image. This scoping decision is made to maximize utilization of maintainer time and attention span on high impact work. 
```